aten/src/TH/generic/THTensor.cpp

```diff
@@ -74,11 +74,12 @@ THTensor *THTensor_(newWithStorage)(THStorage *storage, ptrdiff_t storageOffset,
   if (strides.data()) {
     TORCH_CHECK(sizes.size() == strides.size(), "number of sizes and strides must match");
   }
+  THStorage *new_storage = THStorage_(new)();
   THTensor *self = c10::make_intrusive<at::TensorImpl, at::UndefinedTensorImpl>(
-    c10::intrusive_ptr<at::StorageImpl>::reclaim(THStorage_(new)()),
+    c10::intrusive_ptr<at::StorageImpl>::reclaim(new_storage),
     at::DispatchKey::CPUTensorId
   ).release();
-  THTensor_(setStorageNd)(self, storage, storageOffset, sizes.size(),
+  THTensor_(setStorageNd)(self, storage != nullptr ? storage : new_storage, storageOffset, sizes.size(),
```

Comment:
I know it's not part of the logic change in the PR but I'm educationally curious what's happening here, like why we create a new `THStorage` even if a non-null one is passed. 

Is it that `THStorage` is just a wrapper and so `new_storage` gets the expensive part of `storage` plugged into it if `storage` is non-null?

Reply:
I think it's historical -- storages and tensors used to be more disconnected, so it was possible to e.g. have a tensor without a storage (or to swap the type of a tensor by just replacing its storage).

This kind of forced storages to be first class and had a bunch of corner cases -- i.e. do I need to handle the case where a tensor doesn't have a storage in each of my functions?

Later we decided to disallow those cases, which made the APIs above really convoluted (I need to create a storage because I always need a storage, but I'm just going to immediately wipe out that storage anyway).

So I really think we should just get rid of all these storage-focused APIs.  I _might_ do that in a follow on PR as there aren't many cases left.